### ISSUE TITLE
Item Inits in persistence. 

### ISSUE THREAD

**Issue by TheDragon117** [gruppe-adler/grad-persistence#21]
First I'd like to thank you for all of the work you've done, I love this system and generally its been working fantastically for me. I do have an issue though where I add actions to and when they load back in they will no longer have the action. 

So for example:
```
[ 
 intelobj001,           
 "Hack Tablet",          
 "\a3\ui_f\data\IGUI\Cfg\holdactions\holdAction_connect_ca.paa",  
 "\a3\ui_f\data\IGUI\Cfg\holdactions\holdAction_connect_ca.paa",  
 "_this distance _target < 3",       
 "_caller distance _target < 3",       
 {},              
 {},              
 {[[_target,_caller],"Scripts\phase000_intel001.sqf"] remoteExec ["execVM",2];},     
 {},              
 [],              
 30,              
 0,              
 true,             
 false             
] remoteExec ["BIS_fnc_holdActionAdd", 0, intelobj001];
```

After a save and load action disappears and I can't seem to reference the item using the variable name to add it again on load.

**Comment by Damsous** [comment]
On my side i never try to put an addaction on spawned object but it work for me with FOB box (editor placed), where there is many add action in the init field of the item (vehicle spawner, Ai recruit, Arsenal...)

**Comment by TheDragon117** [comment]
@Damsous after reading your response I did some thinking on it, and you are correct my original statement is inaccurate, an editor placed items init will persist. Spawned objects will lose their actions though. 

It has to be a mistake I'm making on my end on not being able to reference the items I did a test on the same mission where I just created a trigger set to !Alive myobject; and it fired when the object was destroyed so the variable name is persisting. I just can't seem to get the action back on the object after a save.

**Comment by McDiod** [comment]
The problem might be, that the `BIS_fnc_holdActionAdd` function is executed, before the object's variable name is restored (see [Initialization Order](https://community.bistudio.com/wiki/Initialization_Order#targetText=Initialization%20order%20refers%20to%20the,the%20start%20of%20a%20mission.) / *Functions with postInit attribute are called*).

See if this works:

```sqf
[{!isNil "intelobj001" && {!isNull "intelobj001"}},{
    [
        intelobj001,           
        "Hack Tablet",
        ...
    ] call BIS_fnc_holdActionAdd;
},[],60,{diag_log "ERROR: addAction for intelobj001 timed out."}] remoteExecCall ["CBA_fnc_waitUntilAndExecute",0,true];
```

**Comment by TheDragon117** [comment]
So I've tried this a couple of ways. 
I did get an error that `!isNull "intelobj001"` should be `!isNull intelobj001` So I tried it with that correction and by running that in the objects Init, worked when the script was called. but didn't show after save. 

When an intel object is picked up I have it creating a ProfileVariable saying the intel is gone and upon a restart and loading back in it will check in the serverinit to go ahead and run the diary entry that the player should recieve once the intel is found and that works. So I tried  using it in that like:
 ```if (isNil {profileNamespace getVariable Phase000Intel001}) then {
	[{!isNil "intelobj001" && {!isNull "intelobj001"}},{
    [
        intelobj001,           
        "Hack Tablet",
        ...
    ] call BIS_fnc_holdActionAdd;
},[],60,{diag_log "ERROR: addAction for intelobj001 timed out."}] remoteExecCall ["CBA_fnc_waitUntilAndExecute",0,true];
	} else {
	
	phase000_intel001 = 1;

	publicVariable "phase000_intel001";
}; ``` and again the same effect, the action would show when the script was called but not after the mission was saved and reloaded. 

I do appreciate the help in trying to trouble shoot this.

**Comment by TheDragon117** [comment]
I did input the rest of the command in place of the ... And if placed directly on an editor item It does stay even without the 

```
 [{!isNil "intelobj001" && {!isNull "intelobj001"}},{
```

I apologize that it was unclear, it wasn't until I replyied to Damsous that I realized it was only an issue on items spawned with a script that I lose the addactions. So for example my mission starts, and after they've been briefed phase_000 starts  via a publicvariable and excutes a script that spawns several intel items around the map with the addactions  called on them for the players to have to find. 

I spawn them via script since the only editor placed objects I use are player slots and buildings or fixtures that should persist whether destroyed or not. If the intel comes back after being picked up and deleted checks later on to control the mission flow won't work if it exists. Also as there a number of them it obviously isn't great for FPS.

I think the issue is that its added to the object and not run from actual init line being saved in the editor, but I haven't been able to find a way to automatically add the action back to the items after a save/load. When players find the intel items and use the addaction it creates a diary entry with hints on where to look for more, which I've been able to persist by having it trigger a public variable to keep it JIP for the same play session and a profileNameSpace Variable to persist it through a save/load and that works like a charm since diary entries arent persisted.

I think ultimatley to do it I'll need to run something from `initServer.sqf` to check and see if the mission has been saved, if so wait for the save to load and once a saved game has been loaded to check the variable names of those items exist and if so go ahead and add the action back.  to them right after load.

TL;DR I don't think its a hiccup of the persistence script I think it was just I needed to better undestand how it worked and how the method I called things affected it. Though if you have any suggestions as to how I could do the above or if I'm still approaching it wrong tips from guys with more talent and experience then myself are always appreciated.

**Comment by Damsous** [comment]
Why you didn't place the intel in editor ? don't forget to add a variable name.

**Comment by TheDragon117** [comment]
That may be something I have to consider. Usually I try not to have something exist until it needs to, I'm making missions usually with 30+ people on it, sometimes up to 600-700 ai at a given time. This particular mission I had hoped to have around 200 intel items and such. Having an extra 200 items can definitly have an impact on FPS. though I may need to place them in the editor in some hidden place like inside of a large rock or nonenterable building or something to keep them from being found or destroyed when they shouldn't be and, and then just move them where they need to be via script at a later time when they should be able to be found. I just need to adjust everything else to make sure the frames stay at a playable rate.

**Comment by Damsous** [comment]
I think if you use 600 AI they should be flaged as dynamic simulated, just disable simulation on the intel item that will not cause issue for the performance, only the model will be loaded by the client (for sure if the 200 intel are in the same village that will cause a lower framerate client side) but i think the intel are spread on the map
You can also hide/show any object when they are hide its for client/server and can't take damage, so you can do it by sync trigger (server) and show/hide module, or just add this line at the end of your script when you want unhide an intel.
If the intel are hidden don't worry about the performance and it can be done very easely.

for exemple you want hide a intel until the intel_7 is find
put this in the init line of the hidden intel : 
`intel_8 hideObjectGlobal true; // intel_8 is the variable name of the intel object`

add this line at the end of your code when the intel_7 is found :+1: 
`intel_8 hideObjectGlobal false;
deleteVehicle intel_7;`

So the intel_8 will show up when the intel_7 is discovered.

Just an info about the performance, i already do some mission with 500 editor placed AI packed in Kavala by toying with Show/Hide for the unit inside the building or enable dynamic sim for patrol, vehicle or exposed unit... the perf was really great.

For the persistence if the

**Comment by Damsous** [comment]
For the persistence of your diary, just place a trigger for each deleted intel then execute the diary with this trigger, make sure to add a small radius on your trigger with the intel inside.

trigg condition :  (in this exemple the intel is a human skeleton) 
`{_x inArea thisTrigger} count (getPos thisTrigger nearObjects ["Land_HumanSkeleton_F", 25]) == 0`
Activation :  
put your code to create the diary.

So when a mission will be loaded on the server, the trigger will detect the not present intel the create the diary, so you will have a persistence for all the discovered intel.
